aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SetupTOTPCognitoActions.kt

```diff
@@ -0,0 +1,141 @@
+                            logger.verbose("New Session is ${response.session}")
+                            SetupTOTPEvent(
+                                SetupTOTPEvent.EventType.RespondToAuthChallenge(
+                                    eventType.username,
+                                    it.session
+                                )
+                            )
+                        }
+
+                        else -> {
+                            SetupTOTPEvent(
+                                SetupTOTPEvent.EventType.ThrowAuthError(
+                                    Exception("Software token verification failed")
+                                )
+                            )
+                        }
+                    }
+                } ?: SetupTOTPEvent(
+                    SetupTOTPEvent.EventType.ThrowAuthError(Exception("Software token verification failed"))
+                )
+            } catch (exception: Exception) {
+                SetupTOTPEvent(
+                    SetupTOTPEvent.EventType.ThrowAuthError(exception)
+                )
+            }
+            logger.verbose("$id Sending event ${evt.type}")
+            dispatcher.send(evt)
+        }
+
+    override fun respondToAuthChallenge(
+        eventType: SetupTOTPEvent.EventType.RespondToAuthChallenge
+    ): Action =
+        Action<AuthEnvironment>("RespondToAuthChallenge") { id, dispatcher ->
+            logger.verbose("$id Starting execution")
+            val evt = try {
+                val challengeResponses = mutableMapOf<String, String>()
+
+                challengeResponses["USERNAME"] = eventType.username
+                val response = cognitoAuthService.cognitoIdentityProviderClient?.respondToAuthChallenge {
+                    this.session = eventType.session
+                    this.challengeResponses = challengeResponses
+                    challengeName = ChallengeNameType.MfaSetup
+                    clientId = configuration.userPool?.appClient
+                }
+
+                response?.let {
+                    SignInChallengeHelper.evaluateNextStep(
+                        username = eventType.username,
+                        challengeNameType = response.challengeName,
+                        session = response.session,
+                        challengeParameters = response.challengeParameters,
+                        authenticationResult = response.authenticationResult
+                    )
+                } ?: SetupTOTPEvent(
+                    SetupTOTPEvent.EventType.ThrowAuthError(Exception("Software token verification failed"))
+                )
+            } catch (exception: Exception) {
+                SetupTOTPEvent(
+                    SetupTOTPEvent.EventType.ThrowAuthError(exception)
+                )
+            }
+            dispatcher.send(evt)
+        }
+
+    override fun resetToWaitingForAnswer(eventType: SetupTOTPEvent.EventType.ThrowAuthError): Action =
```

Comment:
As removed in challengeSM we may not need this if we are also handling the case in Error state